### DISCUSSION TITLE
Fat jar not including Analog Input?

### DISCUSSION THREAD

**Discussion by Higgins328** [Pi4J/pi4j#205] [Q&A]
Hi,

I have tried the example for building a fat jar using the maven-shade-plugin, and it does include other dependencies I use, but for pi4j I get the error "com.pi4j.provider.exception.ProviderNotFoundException" when I use library com.pi4j.io.gpio.analog.AnalogInput

My code is as follows:

import com.pi4j.Pi4J;
import com.pi4j.io.gpio.analog.AnalogInput;

public class temperaturePiGpio {

    private static final int PIN_1 = 1; // PIN 18 = BCM 24

    public float printTemperature() {

        float retVal = 0.0f;

        System.out.println("GPIO Digital Output ... started.");

        var pi4j = Pi4J.newAutoContext();

        var analogReadConfig = AnalogInput.newConfigBuilder(pi4j)
                .id("aReadConfig")
                .name("aReadConfigName")
                .address(PIN_1);

...

I saw an issue that was closed some time ago where the developer was using a digital input. Should the fat jar approach work also fpor me using the analog input?

**Comment by Higgins328** [comment]
I need to add some more info; I am using a Raspberry pi Zero W running Raspberry Pi OS 32 bit, and Java 11 Azul Jdk. Everything else in the app works fine, I use JavaCV for taking images and some other stuff like FTP to upload images, gRPC to send info to a server. All those dependencies gets apparently included in the jar

**Comment by Higgins328** [comment]
Is this finding recognized as true? If so, are there any plans for fixing this?

**Reply by ronatartifact** [reply]
You might want to show your pom in case there is an issue.
Does it compile and run ok? If so it would seem to be a maven issue.

**Comment by FDelporte** [comment]
As the Raspberry Pi GPIOs are digital-only, analog inputs are indeed not part of Pi4J V2 itself. But there are different solutions possible with additional converter chips. One possible approach with an MCP3008 is explained on https://pi4j.com/featured-projects/joystick-game/

**Reply by Higgins328** [reply]
Aha, thank you for the clarification. Having been using arduinos a bit, that usually have input pins for both digital and analog signals, I assumed that the Raspberry also was capable of reading analog input. My bad! I did a solution that uses an Arduino Nano for reading the actual analog input (in this case a temperature sensor) and the raspberry in turn reads serial data from the Arduino, using the fazecast java library. Works well with RPI 3 and 4 but I cannot get it to work with RPI Zero. Will try to look into MCP3008, or shift to digital temperature sensor instead of analog.
